vyper/venom/parser.py

```diff
@@ -0,0 +1,165 @@
+from vyper.venom.context import IRContext
+from vyper.venom.basicblock import (
+    IRLabel,
+    IRVariable,
+    IRLiteral,
+    IROperand,
+    IRInstruction,
+    IRBasicBlock,
+)
+from vyper.venom.function import IRFunction
+from lark import Lark, Transformer
+
+
+VENOM_PARSER = Lark(
+    """
+    %import common.CNAME
+    %import common.DIGIT
+    %import common.LETTER
+    %import common.WS
+    %import common.INT
+
+    start: function* data_section
+    function: "function" NAME "=>" "{" block* "}"
+    data_section: "data:" call*
+
+    block: NAME ":" (call | assignment)*
+
+    assignment: VAR_IDENT "=" expr
+    expr: call | CONST
+    call: OPCODE operands_list
+
+    operands_list: (operand ("," operand)*)?
+
+    operand: VAR_IDENT | CONST | LABEL
+
+    CONST: INT
+    OPCODE: CNAME
+    VAR_IDENT: "%" INT (":" INT)?
+    LABEL: "@" NAME
+    NAME: (DIGIT|LETTER|"_")+
+
+    %ignore WS
+    """
+)
+
+
+def _set_last_var(fn: IRFunction):
+    for block in fn.get_basic_blocks():
+        output_vars = (
+            instruction.output
+            for instruction in block.instructions
+            if instruction.output is not None
+        )
+        for output_var in output_vars:
+            assert isinstance(output_var, IRVariable)
+            value = output_var.value
+            assert value.startswith("%")
+            fn.last_variable = max(fn.last_variable, int(value[1:]))
+
+
+def _set_last_label(ctx: IRContext):
+    for fn in ctx.functions.values():
+        for bb in fn.get_basic_blocks():
+            label = bb.label.value
+            if label.isdigit():
```

Comment:
this doesn't seem right since labels are usually not just numbers. maybe just set `last_label` to the # of unique seen labels.

Reply:
I mean `last_label` is only used to generate numerical labels so only existing labels that are numerical should affect it, right?

Reply:
i don't think so --
![Screenshot from 2024-12-07 17-47-53](https://github.com/user-attachments/assets/6fe34aa1-e457-4016-bf30-7fe46cf78203)

Reply:
Just setting to the number is kind of brittle though because you might not sequentially number variables for some reason.

Reply:
Ok did the following to handle potential collisions with labels that have a "suffix":

```python
            label_head, *_ = label.split("_", maxsplit=1)
            if label_head.isdigit():
                ctx.last_label = max(int(label_head), ctx.last_label)
```